[IMP] Workspace Symbol request and $/Cancel support

fda-odoo opened the pull request.

mmahrouss commented on server/src/core/file_mgr.rs:
```diff
@@ -397,24 +397,50 @@ impl FileInfo {
     }
 
+    pub fn try_offset_to_position(&self, offset: usize) -> Option<Position> {
+        FileInfo::try_offset_to_position_with_rope(self.file_info_ast.borrow().text_rope.as_ref().expect("no rope provided"), offset)
```
Maybe you want to also not expect here? 

Perhaps:
```suggestion
        FileInfo::try_offset_to_position_with_rope(self.file_info_ast.borrow().text_rope.as_ref()?, offset)
```

mmahrouss commented on server/src/features/workspace_symbols.rs:
```diff
@@ -0,0 +1,186 @@
+        Ok(Some(WorkspaceSymbolResponse::Nested(symbols)))
+    }
+
+    fn browse_symbol(session: &mut SessionInfo, symbol: &Rc<RefCell<Symbol>>, query: &String, parent: Option<String>, parent_path: Option<&String>, can_resolve_location_range: bool, results: &mut Vec<WorkspaceSymbol>) -> bool {
```
Could be nice to explain what the boolean means? 

It is slightly confusing that it returns true if it is cancelled

fda-odoo committed:
[IMP] Workspace Symbol request and $/Cancel support

Result contains classes and functions that contains the query.
It will return model names too, pointing to all the classes implementing it.
An empty query is returning nothing, to prevent too many results
XML_ID are returned too, but prefixed with "xmlid."
